### PR TITLE
Introduce mode `subobjects=auto` for objects

### PR DESCRIPTION
**Pull request by kkrik-es**
The new mode relaxes the behavior of `subobjects=false` to take advantage of any objects that are present in the mappings, and flattens fields otherwise.

Setting `subobjects=false` combines the following:

1. Fields with dotted names are directly created under the object, instead of creating intermediate objects
1. Object mappers are not allowed in the mappings (get flattened)

While the former can lead to indexing errors, as described in https://github.com/elastic/logs-dev/issues/105, the latter is too restrictive and leads to problems with nested objects or using objects with params like `enabled=false` or `store_array_source=true`. 

To that end, the new mode relaxes the behavior of `subobjects=false` and minimizes indexing errors while also allowing for  defining and using objects (including nested) in the mappings.

Related to #108826 and #106812

### CODE REVIEW

**Comment by felixbarny** [comment]
> The new mode relaxes the behavior of subobjects=false to take advantage of any objects that are present in the mappings, and flattens fields otherwise.

To minimize the amount of object mappers even further, we could [auto-flatten](https://www.elastic.co/guide/en/elasticsearch/reference/8.14/subobjects.html#_auto_flattening_object_mappings) object mappers when possible and only add them when auto-flattening fails (for example when `enabled: false` or `dynamic: false` is set or a `nested` field type is used).

**Comment by javanna** [comment]
I gave this some thought and I like the overall direction.

Ideally, subobjects false becomes the future of mappings. That is because objects are mostly placeholders, with no footprint in Lucene, that end up taking quite a bit of space in the heap and causing overhead for no good reason. Declaring objects is mostly an old habit in how mappings are defined, but fields are flattened in Lucene and could rather be flattened in mappings as well.

In order to get there, we can't expect users to adapt their mappings to no longer use objects, hence we introduced auto-flattening of objects. 

Additionally, there are still situations that require using objects (like nested, setting enabled, dynamic etc.). For those, we could remove the need to declare an object to use these features on the long run. On the other hand, if we auto-flatten objects, and keep objects only when strictly necessary, we automatically do what we would like users to do (remove objects when not necessary), without causing pain for users. That sounds like the way to go.

it is important that we do auto-flatten objects, otherwise we open the doors for going back to accepting a lot of needless objects in the mappings. If we do so, it would be great to also keep an eye on how many exceptions we do have to make, as that would help identifying how important it is to decorrelate some of the features attached to object configuration from the object definition on the long run.

**Comment by javanna** [comment]
I have been wondering if there's a possibility that `subobjects:false` gets updated to get the new suggested behaviour. Is that technically a breaking change? We would still auto-flatten objects like we already do, but at the same time, accept objects when they can't be auto-flattened instead of throwing an error.

**Comment by kkrik-es** [comment]
I was also wondering if the current behavior for `sobobjects: false` is a feature or a bug.. I don't think making the `auto` behavior the default for `false` would cause bwc issues, as any mappings with non-conforming objects would be rejected in the past. We'd still need to test upgrading etc very carefully, though, as this would be a much more far-reaching change.

**Comment by felixbarny** [comment]
I think we should go for this separate `auto` mode because changing the semantics for `subobjects: false` may be a breaking change. Setting `subobjects` to `false` gives a strong guarantee that an object mapper doesn't contain any object mappings and that there can't be any object/scalar conflicts. With this new mode, there may still be object scalar conflicts, but they're reduced as much as possible.

However, I do think we should consider using `subobjects: auto` by default in as many places as possible.

> it is important that we do auto-flatten objects, otherwise we open the doors for going back to accepting a lot of needless objects in the mappings.

I agree. Implementation-wise, I think it's fine to separate this, though, as long as both land in the same release.

**Comment by felixbarny** [comment]
> > it is important that we do auto-flatten objects, otherwise we open the doors for going back to accepting a lot of needless objects in the mappings.
>
> I agree. Implementation-wise, I think it's fine to separate this, though, as long as both land in the same release.

I saw this was marked as ready to review but I think the above suggestion to auto-flatten mappings whenever possible wasn't implemented, yet.

Are you intending to follow up with another PR that adds this behavior and then also add some user-facing docs for this new `subobjects` mode?

Also, I'm thinking that it would be useful for `passthrough` fields to use this mode rather than not allowing any subobjects at all. That would allow passthrough fields to support `nested` fields, for example, which would be useful for OTel attributes.

**Review by lkts — APPROVED** [review]
I have to admit it's hard to follow. Parts that i could understand LGTM. `StandardVersusLogsIndexModeRandomDataChallengeRestIT` should provide decent coverage for this.

File: server/src/main/java/org/elasticsearch/index/mapper/NestedObjectMapper.java [hunk 1]
```diff
@@ -441,7 +442,7 @@ public DocValuesLoader docValuesLoader(LeafReader leafReader, int[] docIdsInLeaf
 
         private List<Integer> collectChildren(int parentDoc, BitSet parentDocs, DocIdSetIterator childIt) throws IOException {
             assert parentDocs.get(parentDoc) : "wrong context, doc " + parentDoc + " is not a parent of " + nestedTypePath;
-            final int prevParentDoc = parentDocs.prevSetBit(parentDoc - 1);
+            final int prevParentDoc = parentDoc > 0 ? parentDocs.prevSetBit(parentDoc - 1) : -1;
```
[the remark this answers is not in the document]
Comment (kkrik-es) [reply server/src/main/java/org/elasticsearch/index/mapper/NestedObjectMapper.java:445]: https://github.com/elastic/elasticsearch/issues/111992 ..

**Comment by felixbarny** [comment]
Makes sense.  I've created https://github.com/elastic/elasticsearch/issues/112013 to discuss changes for `passthrough`.

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/index/92_metrics_auto_subobjects.yml [hunk 2]
```diff
@@ -0,0 +1,250 @@
+---
+"Metrics object indexing":
+  - requires:
+      test_runner_features: allowed_warnings_regex
+      cluster_features: ["gte_v8.3.0", "mapper.subobjects_auto"]
```
[the remark this answers is not in the document]
Comment (kkrik-es) [reply rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/index/92_metrics_auto_subobjects.yml]: Done.

File: server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java [hunk 3]
```diff
@@ -159,7 +206,7 @@ private static ObjectMapper.Builder findObjectBuilder(String fullName, DocumentP
             if (objectMapper != null) {
                 return objectMapper.newBuilder(context.indexSettings().getIndexVersionCreated());
             }
-            throw new IllegalStateException("Missing intermediate object " + fullName);
+            return null;
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:209]: Maybe comment here what returning null means here?

File: server/src/test/java/org/elasticsearch/index/mapper/DynamicTemplatesTests.java [hunk 4]
```diff
@@ -1434,6 +1442,248 @@ public void testSubobjectsFalseWithInnerNestedFromDynamicTemplate() {
...
+        }));
+
+        DocumentParsingException err = expectThrows(DocumentParsingException.class, () -> mapper.parse(source("""
+            {
+              "object" : [
+                {}
+              ]
+            }
+            """)));
+        assertEquals("[3:5] Tried to add nested object [object] to object [_doc] which does not support subobjects", err.getMessage());
```
Comment (martijnvg) [inline server/src/test/java/org/elasticsearch/index/mapper/DynamicTemplatesTests.java:1573]: In the context of enabling subobjects=auto by default for logsdb, does this add another barrier? I don't think we fail today in this situation because subobjects are enabled?

**Review by martijnvg — APPROVED** [review]
LGTM, thanks for working on this Kostas!

I left on question around enabling auto subsubject by default.

File: server/src/test/java/org/elasticsearch/index/mapper/DynamicTemplatesTests.java (hunk 4 above)
Comment (kkrik-es) [reply server/src/test/java/org/elasticsearch/index/mapper/DynamicTemplatesTests.java:1573]: Good catch, this actually should work. Updated.
